app/services/user_access_key.rb

```diff
@@ -76,4 +85,18 @@ def z1_with_padding(len)
     end
     str
   end
+
+  def build(password, salt)
+    scrypt_salt = COST + OpenSSL::Digest::SHA256.hexdigest(salt)
+    scrypted = SCrypt::Engine.hash_secret password, scrypt_salt, 32
+    segment_one, segment_two = build_segments(scrypted)
+    random_r = Pii::Cipher.random_key
+    { salt: scrypt_salt, z_one: segment_one, z_two: segment_two, random_r: random_r }
+  end
+
+  def build_segments(scrypted)
+    hashed = SCrypt::Password.new(scrypted).digest
+    segment_one = hashed.slice!(0...32)
+    [segment_one, hashed]
+  end
```

Comment:
Sorry. I was viewing a single commit instead of all the changes and didn't get the whole context.